{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPWqaczld+zm+TPb0RVKTte"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**WEEK 5**\n","\n","This week we'll be working through Chapter 3 of Applied Bioinformatics. We will be running BLAST on the command line after reading week, and this textbook has a really good explanation of how that is implemented. For today's lab, we'll be looking at how sequence alignment works under the hood using the two alignment methods we discussed in class in Week 3: global alignment and local alignment.\n","\n","First, we have to install BioPython again, since this is a new notebook. You can do that by running the code block below."],"metadata":{"id":"Xq0nuLZyj_ZT"}},{"cell_type":"code","source":["##Install BioPython in Jupyter Notebook\n","%pip install biopython"],"metadata":{"id":"USgey0rGkxzI"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Before we start, try having a go at aligning some sequences using the Hamming distance between them.\n","\n"],"metadata":{"id":"qs_6QkK11o1-"}},{"cell_type":"code","source":["seq_1 = \"ATCGATCGATCG\"\n","seq_2 = \"ATCGATCGATCA\"\n","distance = sum(n1 != n2 for n1, n2 in zip(seq_1, seq_2))\n","\n","print(\"The Hamming distance between these two sequences is \" + str(distance))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"20gFMcHf3hTe","executionInfo":{"status":"ok","timestamp":1702952920822,"user_tz":420,"elapsed":355,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"9ef40e06-56ca-4df5-d93d-5628e9d399f4"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["The Hamming distance between these two sequences is 1\n"]}]},{"cell_type":"markdown","source":["Try changing the sequences. What kinds of sequences can be aligned? What sequences can't be aligned?"],"metadata":{"id":"NPbN_Tqi54Op"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"BKVfygtx5_aN"}},{"cell_type":"markdown","source":["#Part 1: Global Alignment\n","Recall what you have learned about the Needleman-Wunsch algorithm, which we covered in class about two weeks ago. Well, you covered it in class two weeks ago, I while writing this in December 2023 haven't even finished prepping that lecture yet. Time is weird.\n","\n","In the lecture, we used the pre-computed version of the algorithm written by the textbook author and implemented in numpy. As you can imagine, BioPython has a handily written function that does this job for us as well. In the previous iteration of BioPython one could specify which algorithm to use; with the new align package the choice of global alignment algorithm is optimised behind the scenes."],"metadata":{"id":"IzLKE6vskyVO"}},{"cell_type":"code","source":["##This code defines two sequences and aligns them using global alignment.\n","from Bio import Align\n","aligner = Align.PairwiseAligner()\n","aligner.mode = 'global'\n","aligner.match_score = 2\n","aligner.mismatch_score = -1\n","alignment = aligner.align(\"TACTAA\", \"ATCTG\")\n","for alignment in aligner.align(\"TACTAA\", \"ATCTG\"):\n","    print(\"Score = %.1f:\" % alignment.score)\n","    print(alignment)"],"metadata":{"id":"1dcIcpPioPiq"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Is this the output you expected? Why or why not?"],"metadata":{"id":"vAkb0QEquNzJ"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"ryU2UZOjqGI7"}},{"cell_type":"markdown","source":["Try altering the `match_score` and `mismatch_score` parameters. How does this affect the output? Why do you think that is?"],"metadata":{"id":"zMTppCYZrkkx"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"Jslpa0GYr7pl"}},{"cell_type":"markdown","source":["#Part 2: Local Alignment\n","Let's repeat the steps above, but with a local instead of global algorithm."],"metadata":{"id":"hJ9Z4ojgWfNn"}},{"cell_type":"code","source":["##This code defines two sequences and aligns them using global alignment.\n","from Bio import Align\n","aligner = Align.PairwiseAligner()\n","aligner.mode = 'local'\n","aligner.match_score = 2\n","aligner.mismatch_score = -1\n","alignment = aligner.align(\"TACTAA\", \"ATCTG\")\n","for alignment in aligner.align(\"TACTAA\", \"ATCTG\"):\n","    print(\"Score = %.1f:\" % alignment.score)\n","    print(alignment)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IHq-o6p7zPtp","executionInfo":{"status":"ok","timestamp":1702951365757,"user_tz":420,"elapsed":3,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"086e5687-1207-4756-8b89-57100e56ac5e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Score = 6.0:\n","target            1 A-CT 4\n","                  0 |-|| 4\n","query             0 ATCT 4\n","\n","Score = 6.0:\n","target            0 TACT 4\n","                  0 |-|| 4\n","query             1 T-CT 4\n","\n"]}]},{"cell_type":"markdown","source":["Is this the output you expected? Why or why not?"],"metadata":{"id":"fgZYKbf4zmR4"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"qgavdG8Jzr7j"}},{"cell_type":"markdown","source":["Try altering the `match_score` and `mismatch_score` parameters. How does this affect the output? Why do you think that is?"],"metadata":{"id":"h35FzkLXzr7j"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"uNxFGbspzr7k"}},{"cell_type":"markdown","source":["#Part 3: Substitution matricies\n","Substitution matricies allow us to add weight to our alignment calculations. We've already sort of been doing that with the `match_score` and the `mismatch_score`, but substitution matricies allow us to define these relationships in more detail.\n","\n","First, let's define a simple substitution matrix: a score of 2 for no change, a score of 0 for a transtition (pyrimidine - pyrimidine or purine - purine) and -1 for a transversion (pyrimidine - purine or vice versa)\n"],"metadata":{"id":"wEzOuSj43lwq"}},{"cell_type":"code","source":["##This code defines a substitution matrix\n","substitution_matrix = {\n","    'A': {'A': 2, 'C': -1, 'G': 0, 'T': -1},\n","    'C': {'A': -1, 'C': 2, 'G': -1, 'T': 0},\n","    'G': {'A': 0, 'C': -1, 'G': 2, 'T': -1},\n","    'T': {'A': -1, 'C': 0, 'G': -1, 'T': 2}\n","}"],"metadata":{"id":"A8TX9YU5DwEr"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["How many times does each number appear in the matrix? Why?"],"metadata":{"id":"K5232n0LE11O"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"DHORwyeZE45t"}},{"cell_type":"markdown","source":["To actually *use* our substitution matrix, we will need to define a function that allows us to pull scores from it. Don't worry too much about how this is done as Python functions are beyond the scope of this course, but feel free to ask Brian annoying questions about how it works. The important thing to know is that when you define a function using `def`, you set **parameters**, which are the pieces of information the function needs to know to run. For this function, which calculates the substitution score between each amino acid, only two parameters are needed: the original base pair and the substituted base pair."],"metadata":{"id":"RuovmQUQECiR"}},{"cell_type":"code","source":["##This code defines a function that calculates substitution scores and shows a test example\n","def get_substitution_score(nucleotide1, nucleotide2):\n","    return substitution_matrix[nucleotide1][nucleotide2]\n","\n","score = get_substitution_score(\"A\",\"C\")\n","print(score)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"EQvfhVzQEWls","executionInfo":{"status":"ok","timestamp":1702956486265,"user_tz":420,"elapsed":411,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"1147f7f0-ac5f-452d-8d25-ce1c470929b5"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["-1\n"]}]},{"cell_type":"markdown","source":["Now we're going to combine the Needleman-Wunsch algorithm (one of our global alignment algorithms defined above) with our substitution matrix to run some alignments. The code in this block is DEFINITELY beyond the scope of this course, but once again feel free to ask Brian annoying questions about it."],"metadata":{"id":"6ILUOMBqFV82"}},{"cell_type":"code","source":["def needleman_wunsch(sequence1, sequence2, substitution_matrix, gap_penalty=-1):\n","    # Initialize the scoring matrix\n","    rows = len(sequence1) + 1\n","    cols = len(sequence2) + 1\n","    score_matrix = [[0] * cols for _ in range(rows)]\n","\n","    # Initialize the scoring matrix with gap penalties\n","    for i in range(1, rows):\n","        score_matrix[i][0] = score_matrix[i-1][0] + gap_penalty\n","\n","    for j in range(1, cols):\n","        score_matrix[0][j] = score_matrix[0][j-1] + gap_penalty\n","\n","    # Fill in the scoring matrix based on the substitution matrix\n","    for i in range(1, rows):\n","        for j in range(1, cols):\n","            match_score = score_matrix[i-1][j-1] + substitution_matrix[sequence1[i-1]][sequence2[j-1]]\n","            gap1_score = score_matrix[i-1][j] + gap_penalty\n","            gap2_score = score_matrix[i][j-1] + gap_penalty\n","\n","            # Choose the maximum score at the current position\n","            score_matrix[i][j] = max(match_score, gap1_score, gap2_score)\n","\n","    # Traceback to find the optimal alignment\n","    alignment1 = \"\"\n","    alignment2 = \"\"\n","    i, j = rows - 1, cols - 1\n","\n","    while i > 0 or j > 0:\n","        if i > 0 and j > 0 and score_matrix[i][j] == score_matrix[i-1][j-1] + substitution_matrix[sequence1[i-1]][sequence2[j-1]]:\n","            alignment1 = sequence1[i-1] + alignment1\n","            alignment2 = sequence2[j-1] + alignment2\n","            i -= 1\n","            j -= 1\n","        elif i > 0 and score_matrix[i][j] == score_matrix[i-1][j] + gap_penalty:\n","            alignment1 = sequence1[i-1] + alignment1\n","            alignment2 = \"-\" + alignment2\n","            i -= 1\n","        else:\n","            alignment1 = \"-\" + alignment1\n","            alignment2 = sequence2[j-1] + alignment2\n","            j -= 1\n","\n","    return alignment1, alignment2"],"metadata":{"id":"jIfSvfolFw1d"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Which parameters are included in the function? What does each of them mean?"],"metadata":{"id":"sqXnUJIkGUac"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"3arWJk1JGYSP"}},{"cell_type":"markdown","source":["Now we have a function and a substitution matrix; let's try running it!"],"metadata":{"id":"fftwKM7ZGcEZ"}},{"cell_type":"code","source":["##This code defines two sequences and runs a Needleman-Wunsch alignment of those sequences using a custom substitution matrix.\n","\n","seq_1 = \"Your DNA sequence\"\n","seq_2 = \"Your DNA sequence\"\n","\n","alignment = needleman_wunsch(seq_1, seq_2, substitution_matrix)\n","\n","print(alignment)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"IpchDZgZGiWL","executionInfo":{"status":"ok","timestamp":1702956474967,"user_tz":420,"elapsed":362,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"07cf9ab8-4223-4c20-ed0b-172f39247ecc"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["('ATGTGTTAGAT', 'ATGTGTTAG-T')\n"]}]},{"cell_type":"markdown","source":["Have a go at changing the sequences. What do you observe?"],"metadata":{"id":"XWcDayMJHt0F"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"HaMDnEPiH0Ee"}},{"cell_type":"markdown","source":["Â© Elisabeth Richardson, 2023. Adapted from Applied Bioinformatics by David A. Hendricks under a [CC-by-4.0 license](https://https://creativecommons.org/licenses/by/4.0/) and shared under same."],"metadata":{"id":"keHeH_QMTkm9"}}]}