{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOLgWFt1Lm/UMsXxn0bQn3m"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**WEEK 3**\n","\n","This week we'll be working through an adapted version of Exercises 4-5 and Lab 2 from Applied Bioinformatics.\n","\n","First, we have to install BioPython again, since this is a new notebook. You can do that by running the code block below."],"metadata":{"id":"FJ0sj8PtSwZc"}},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"HZr77PoaSuRp","executionInfo":{"status":"ok","timestamp":1701466716321,"user_tz":420,"elapsed":10427,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"8bf84dae-9c00-45a9-c8c8-df11556e7052"},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m29.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.23.5)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.81\n"]}],"source":["##Install BioPython in Jupyter Notebook\n","%pip install biopython"]},{"cell_type":"markdown","source":["#Part 1: FastQ\n","First, make sure that you've read Section 1.3.2 of the textbook (FastQ). If you haven't read it, do that now.\n","\n","We're going to look at some fastq data, but before that we need to download it from the Applied Bioinformatics teaching server. We're going to do that using wget, which is a command-line utility. In Jupyter notebooks, you run it using the following command:\n","\n","\n","\n","```\n","!wget <full link to resource>\n","```\n","\n","The URL for the data file we want is:\n","\n","\n","\n","```\n","https://hendrixlab.cgrb.oregonstate.edu/teaching/reads.fastq\n","```\n","\n","Try running the code block below, then check your directory.\n"],"metadata":{"id":"ic_hrzeWVpAN"}},{"cell_type":"code","source":["!wget https://hendrixlab.cgrb.oregonstate.edu/teaching/reads.fastq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"NVDJEsduWVDy","executionInfo":{"status":"ok","timestamp":1701470193356,"user_tz":420,"elapsed":1169,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"027f5517-fb21-4592-b4e1-c7c749c0974d"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["--2023-12-01 22:37:54--  https://hendrixlab.cgrb.oregonstate.edu/teaching/reads.fastq\n","Resolving hendrixlab.cgrb.oregonstate.edu (hendrixlab.cgrb.oregonstate.edu)... 128.193.83.38\n","Connecting to hendrixlab.cgrb.oregonstate.edu (hendrixlab.cgrb.oregonstate.edu)|128.193.83.38|:443... connected.\n","HTTP request sent, awaiting response... 200 OK\n","Length: 2595548 (2.5M)\n","Saving to: ‘reads.fastq’\n","\n","reads.fastq         100%[===================>]   2.47M  3.90MB/s    in 0.6s    \n","\n","2023-12-01 22:37:55 (3.90 MB/s) - ‘reads.fastq’ saved [2595548/2595548]\n","\n"]}]},{"cell_type":"markdown","source":["Now let's import your data as a variable so we can work with it in BioPython."],"metadata":{"id":"y359sY8zhK2r"}},{"cell_type":"code","source":["from Bio import SeqIO\n","fastqFile = \"reads.fastq\" # FASTQ file name\n","data = SeqIO.parse(fastqFile,\"fastq\") # parse the FASTQ file"],"metadata":{"id":"8ACZeW5Sho1Z"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The mechanism for how Phred scores are calculated is discussed in detail in the textbook. If you'd like to try running any of the commands in this section, you can use the code block below. Make sure you understand what a Phred score is - this will come up again later!"],"metadata":{"id":"hUIYFqPRjorq"}},{"cell_type":"code","source":["##Your code here. Try pasting in some commands from the textbook if you want to see what they do in practice."],"metadata":{"id":"3pxeh7wTkTyv"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now it's time for Exercise 4 (with some extra bits)\n","\n",">1. What probability of error corresponds to the following quality score characters:\n",">\n",">- !\n",">- A\n",">- a\n",">- %\n",">\n",">Are these good or bad scores? How do you know?\n","\n","2. Do Phred scores get better or worse as a sequence extends? How does this correspond to the likelihood of error?"],"metadata":{"id":"OjaiNNVVWVuE"}},{"cell_type":"markdown","source":["You can use this text box to write your answers."],"metadata":{"id":"cr5lInRfV62L"}},{"cell_type":"markdown","source":["#Part 2: Applied Bioinformatics Lab 2\n","\n","First, use the wget command to download the following script:\n","\n","````https://hendrixlab.cgrb.oregonstate.edu/teaching/printAverageQualityScores.py````\n"],"metadata":{"id":"sY_u9OAte3ui"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"YCk1BqVDkxVK"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Next, let's count the number of lines in the fastq file. Given the result, how many reads do you think are contained in this file? Why?"],"metadata":{"id":"WPvDsmnZk4i_"}},{"cell_type":"code","source":["!wc -l reads.fastq"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"gHer6yoElAFT","executionInfo":{"status":"ok","timestamp":1701471078722,"user_tz":420,"elapsed":235,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"e44ea5a0-a51d-4888-ff97-a07ac3a03fae"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["40000 reads.fastq\n"]}]},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"XJ06uGX5moIP"}},{"cell_type":"markdown","source":["Next, let's check what happens when you run the script without any input."],"metadata":{"id":"thccQe44lFYg"}},{"cell_type":"code","source":["!python printAverageQualityScores.py"],"metadata":{"id":"tQHgYLnJlIX7"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Looks like something's missing: let's add the reads.fastq file and see what happens."],"metadata":{"id":"5I4_tuK3lij3"}},{"cell_type":"code","source":["!python printAverageQualityScores.py reads.fastq"],"metadata":{"id":"aWdPbgfHmG4a"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The script will create a file called quality.pdf, which is saved in your directory. How does this plot look? What does this say about errors in sequencing data?"],"metadata":{"id":"m_7DzCUpmUfl"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"gInnEBLbmbph"}},{"cell_type":"markdown","source":["We're going to leave the Lab 2 tutorial here, but there are many tools one can use to trim a FastQ file based on quality parameters. My personal favourite is FastQC, which you can learn more about [here.](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)\n","\n","Compare the report outputs from \"Good Illumina Data\" and \"Bad Illumina data\". in the Sample Reports section. For each of the eleven panels, compare and contrast the results. How does the output tell you one is a \"good\" fastq and one is a \"bad\" fastq?"],"metadata":{"id":"xPM73RUFnTrV"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"y5MQ5qhVoFms"}},{"cell_type":"markdown","source":["© Elisabeth Richardson, 2023. Adapted from Applied Bioinformatics by David A. Hendricks under a [CC-by-4.0 license](https://https://creativecommons.org/licenses/by/4.0/) and shared under same."],"metadata":{"id":"cquJ2rgkUSBK"}}]}