{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyPqhNjIGxGDwc40i1qFactF"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**WEEK 7**\n","\n","In this lab, we'll be returning to Chapter 4 of the Applied Bioinformatics textbook, and we'll be looking at the chapter on multiple sequence alignments / phylogenetics.\n","\n","First, since this is a new notebook we'll have to install BioPython using the code block below."],"metadata":{"id":"eANewadiUFak"}},{"cell_type":"code","source":["##Install BioPython in Jupyter Notebook\n","%pip install biopython"],"metadata":{"id":"H0V0PwlW9054"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["##Part 1: Multiple sequence alignment\n","\n","Multiple sequence alignment involves scaling up the concept of pairwise alignment; that is instead of aligning two sequences against each other, you align three or more sequences against each other.\n","\n","If you haven't read Section 4.1 of the textbook, do it now.\n","\n","You'll note that they discuss four methods of multiple sequence alignment; dynamic programming, progressive alignment, iterative alignment, and genome alignment. The categorization of MFA algorithms is kind of blurry and this is one categorization among many; the main ones we talked about in class are pairwise vs iterative, but there's many algorithms which use an adapted version of either approach or a combination of the two. The most important things to remember for now are:\n","\n","1. Multiple sequence alignment is computationally difficult\n","\n","2. All MSA algorithms have to balance speed vs complexity. The more complex the algorithm the more likely they are to be accurate, but the slower they tend to be.\n","\n","Up until this point we've been using BioPython for all our bioinformatic analyses: you might be asking at this point whether you can use BioPython to carry out a multiple sequence alignment. Of course you can! However, it is a bit more complicated than you might think."],"metadata":{"id":"mUu2DMTEWKmy"}},{"cell_type":"code","source":["##Align sequences using the ClustalW algorithm implemented in BioPython\n","from Bio.Align.Applications import ClustalwCommandline\n","\n","\n","in_file = \"example_sequences.fasta\"\n","\n","clustalw_cline = ClustalwCommandline(\"clustalw2\", infile=\"example_sequences.fasta\")\n","\n","print(clustalw_cline)\n","\n","clustalw2 -infile=example_sequences.fasta\n","\n","from Bio import AlignIO\n","align = AlignIO.read(\"example_sequences.aln\", \"clustal\")\n","print(align)"],"metadata":{"id":"JTphpRUKvJQ6"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["The code above *should* perform a MSA using the ClustalW algorithm, but it doesn't. That's because BioPython doesn't directly contain the code necessary to run the ClustalW algorithm; it has what's called a **wrapper**, which allows you to implement the algorithm without having to exit the coding environment you're currently using. However, this assumes that there is something inside the wrapper. In this case, the BioPython module is looking for the ClustalW software inside the wrapper, and is finding... nothing!\n","\n","Instead of installing ClustalW from source in a Jupyter notebook, we're going to use the online server where ClustalW is handily installed for you. You can find the online version [here.](https://https://www.genome.jp/tools-bin/clustalw)"],"metadata":{"id":"YB8lFLyd5oFN"}},{"cell_type":"markdown","source":["**Try using the Clustal online server to align the following sequences, and paste them into the text box below. Make sure you set the parameters correctly and save the output as a FASTA sequence. Include both the clustalw.fasta section and the clustalw.dnd section**\n","\n","```\n",">Sequence_1\n","ATCGATCGATCGAATCGATCG\n",">Sequence_2\n","ATCGATATCGATCGATCGATCG\n",">Sequence_3\n","ATCGATCGATCGATCGACCTCG\n","```\n","\n"],"metadata":{"id":"sujLpOJ46uXN"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"FQzFVMgE77UD"}},{"cell_type":"markdown","source":["**What do you notice about the order of the sequences in your alignment? Why do you think it might be? (Hint: ClustalW is a *progressive* alignment algorithm)**"],"metadata":{"id":"MBj9nd5m7-Tl"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"eSaHUV618JZ8"}},{"cell_type":"markdown","source":["**How might this be different if ClustalW was an *iterative* alignment algorithm?**"],"metadata":{"id":"JkY7AojN-Fgx"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"NG3q8sJv-Q2W"}},{"cell_type":"markdown","source":["##Part 2: Tree Building\n","Tree building involves visualization of multiple sequence alignments, including the quantification of similarity / probability that this is the best alignment as node values / branch lengths. When we think of trees, we think of visual images with branches and leaves and nodes, but computers don't care about those. Trees are often stored in Newick format, which is much more compact and efficient (but much less readable for us poor humans!)\n","\n","The textbook contains an example of a tree written in Newick format, which we will reproduce below (with some minor edits to make it work in Jupyter format, you're welcome). Before we start, make sure you've read Section 4.2.1 of the textbook."],"metadata":{"id":"JuBzDE1y8XSK"}},{"cell_type":"code","source":["##Import the relevant packages\n","from Bio import Phylo\n","\n","##Save example as treefile\n","tree_content = \"((A,B),(C,D));\"\n","output_file = \"tree.txt\"\n","\n","with open(output_file, \"w\") as tree_file:\n","    tree_file.write(tree_content)\n","\n","##Visualise tree\n","tree = Phylo.read(\"tree.txt\",\"newick\")\n","Phylo.draw_ascii(tree)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rOS5UFtn9kGr","executionInfo":{"status":"ok","timestamp":1703803321110,"user_tz":420,"elapsed":185,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"a7532968-b09a-4608-9c61-ea6435204b4b"},"execution_count":47,"outputs":[{"output_type":"stream","name":"stdout","text":["                                        _____________________________________ A\n","  _____________________________________|\n"," |                                     |_____________________________________ B\n","_|\n"," |                                      _____________________________________ C\n"," |_____________________________________|\n","                                       |_____________________________________ D\n","\n"]}]},{"cell_type":"markdown","source":["**Run the code block. What do you notice about the tree?**"],"metadata":{"id":"8psnJTM4--5N"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"jSd5Sx4n_IDw"}},{"cell_type":"markdown","source":["This is a great representation of how the sequences are related, but it doesn't include any information on weighting. Weighting is a way of drawing a tree that takes into account how similar two sequences are; this is why two sequences that are close together have short branches and two sequences that are very different have long branches. Newick format can also take this information into account.\n","\n","There is some information on pairwise distance calculation in Section 4.2 of the textbook, but huge plot twist: we have actually ALREADY generated a weighted tree in Newick format! Look up above where you pasted the clustalw.dnd output - is it starting to look familiar? We're going to use the code block below to visualize it."],"metadata":{"id":"w62SHT9Q_JNS"}},{"cell_type":"code","source":["##Import the relevant packages\n","from Bio import Phylo\n","\n","##Save example as treefile\n","newick_tree_string = \"(\\n  Sequence_1:0.24459,\\n  Sequence_2:0.13636,\\n  Sequence_3:0.13636\\n);\"\n","output_file = \"weighted_tree.txt\"\n","\n","with open(output_file, \"w\") as tree_file:\n","    tree_file.write(newick_tree_string)\n","\n","##Visualise tree\n","tree = Phylo.read(\"weighted_tree.txt\",\"newick\")\n","Phylo.draw_ascii(tree)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"8mPqFFs2A5Uv","executionInfo":{"status":"ok","timestamp":1703794297166,"user_tz":420,"elapsed":374,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"8e612f18-8dd3-4398-a7d9-0a2b6e90b4fb"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["  __________________________________________________________________ Sequence_1\n"," |\n","_|____________________________________ Sequence_2\n"," |\n"," |____________________________________ Sequence_3\n","\n"]}]},{"cell_type":"markdown","source":["**What is the name for a node with more than two branches? What type of tree cannot have these nodes? Which sequence has the longest branch, and what does that mean?**"],"metadata":{"id":"_v55hOpjROeq"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"JeVmSIJNRV-z"}},{"cell_type":"markdown","source":["To establish branch lengths, you have to have a substitution model of evolution; basically, a set of *assumptions* you are willing to work under about how quickly a given site can be expected to change (these are sort of like a substitution matrix, but they include time as a variable). These are described in a lot more detail than you need to know in Section 4.3.2; the most important thing to remember here is that before we build a tree we have to select a substitution model, and there are programmes available (or integrated into the software you use like IQ-TREE2) that can help you select the best model."],"metadata":{"id":"p4x2RrgdRWsC"}},{"cell_type":"markdown","source":["##Part 3: Tree Searching and Evaluating Quality of a Phylogenetic Tree\n","\n","So now we've figured out how to make a multiple sequence alignment and visualise / score a single phylogenetic tree. However, in real life we're usually working with many sequences, and very complex evolutionary relationships. There might be *many* potential trees, and very subtle differences between them which aren't always captured by that initial alignment and guide tree.\n","\n","The solution to this problem is pretty simple: generate lots of very similar trees and evaluate all of them (this is also called \"searching the tree space\"). Then you use statistics to determine which of the trees you generate is most likely to be the correct one.\n","\n","The mechanisms for generating the many similar trees are described in 4.3.10.\n"],"metadata":{"id":"PNxyCAObS3kl"}},{"cell_type":"markdown","source":["**What two mechanisms are used to generate similar trees to the original tree?**"],"metadata":{"id":"GRZAYih0Z1Nn"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"Fwz03SMRaDUB"}},{"cell_type":"markdown","source":["These methods generate trees. We also need a way of evaluating which is the best tree within these tree spaces, and this is where our two statistical methods from the pre-Reading Week lectures come in: Maximum Likehood and Bayesian.\n","\n","We're going to generate a Maximum Likelihood and Bayesian tree and compare them using online servers, but first we need some sequence data.\n","```\n",">Protein_1\n","MAGHTGFTYILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRL\n",">Protein_2\n","MAGHTGFTYILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRL\n",">Protein_3\n","MAGHTGFTYILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRL\n",">Protein_4\n","ILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRLMAGHTGF\n",">Protein_5\n","ILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRLMAGHTGF\n",">Protein_6\n","ILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRLMAGHTGF\n",">Protein_7\n","MAGHTGFTYILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRL\n",">Protein_8\n","MAGHTGFTYILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRL\n",">Protein_9\n","ILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRLMAGHTGF\n",">Protein_10\n","ILGKPVDKVVECTEKACVNELSQRAVTWLGNHIDPADPTVEVRGGHLTECQRELGDKANFTLRNIPFIMLRLHQPDMVAQIAAKQKWGRLMAGHTGF\n","```\n","You can probably tell just by looking at this dataset which sequences are most likely to be closely related. Let's see if your assumptions are correct!\n","\n","\n"],"metadata":{"id":"cvl4HCR4aFkH"}},{"cell_type":"markdown","source":["We're going to use the IQ-TREE2 online server to generate a maximum likelihood tree from this data. Click [here](http://iqtree.cibiv.univie.ac.at/) to access it. It only takes alignment files, so first you'll have to align the above sequences, save them as a file and upload them to the server in FASTA format. Luckily, we just learned a method to do just that.\n","\n","Once you have your file, upload it to the relevant box and leave the rest of the settings alone (this is also called *running with default parameters*). This might take a few minutes, but once it's complete you can access by clicking the \"Analysis results\" tab.\n","\n","What does the resulting tree look like? Use the code block below to visualize the consensus tree."],"metadata":{"id":"oRVuYnsjaqK3"}},{"cell_type":"code","source":["##Import the relevant packages\n","from Bio import Phylo\n","\n","##Save example as treefile\n","newick_tree_string = \"<paste your code here>\"\n","output_file = \"weighted_tree.txt\"\n","\n","with open(output_file, \"w\") as tree_file:\n","    tree_file.write(newick_tree_string)\n","\n","##Visualise tree\n","tree = Phylo.read(\"weighted_tree.txt\",\"newick\")\n","Phylo.draw_ascii(tree)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"-ibdNpOddBD5","executionInfo":{"status":"ok","timestamp":1703801618487,"user_tz":420,"elapsed":194,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"25a4d17d-29bf-4c42-ac6c-c85a8c270236"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":[" , Protein_9\n"," |\n"," |                                           _____________________ Protein_10\n"," |                                          |\n"," |                                          |                     , Protein_1\n"," |                     _____________________|                     |\n"," |                    |                     |                     | Protein_8\n"," |                    |                     |                     |\n"," |____________________|                     |_____________________| Protein_7\n","_|                    |                                           |\n"," |                    |                                           | Protein_3\n"," |                    |\n"," |                    |_____________________ Protein_2\n"," |\n"," | Protein_4\n"," |\n"," | Protein_5\n"," |\n"," | Protein_6\n","\n"]}]},{"cell_type":"markdown","source":["**What is a consensus tree? What does it represent?**"],"metadata":{"id":"nt5fnVT_cvaT"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"DcvAT-lndVhk"}},{"cell_type":"markdown","source":["One of the reasons we tend to use maximum likelihood methods first is because they run a LOT more quickly than Bayesian methods. Because of this (and because Bayesian servers that are free, online, and don't crash every five minutes are few and far between) I have already run the equivalent analysis using a Bayesian method. The resulting Newick file is saved in the notebook as nwktree.txt.\n","\n","Visualize the Newick file."],"metadata":{"id":"nbqMGo8RdWOG"}},{"cell_type":"code","source":["##You can put your code in this block.\n","tree = Phylo.read(\"nwktree.txt\",\"newick\")\n","Phylo.draw_ascii(tree)"],"metadata":{"id":"tFSeB3vKgb0h","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1703803390406,"user_tz":420,"elapsed":167,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"5aafa8d4-177c-4629-fc16-768a417608f8"},"execution_count":48,"outputs":[{"output_type":"stream","name":"stdout","text":["  ____________ Protein_9\n"," |\n"," |             ____________ Protein_10\n"," |____________|\n"," |            |____________ Protein_7\n"," |\n"," |                          ____________ Protein_6\n"," |             ____________|\n","_|            |            |             ____________ Protein_2\n"," |            |            |____________|\n"," |____________|                         |             ____________ Protein_3\n"," |            |                         |____________|\n"," |            |                                      |____________ Protein_8\n"," |            |\n"," |            |____________ Protein_1\n"," |\n"," |             ____________ Protein_4\n"," |____________|\n","              |____________ Protein_5\n","\n"]}]},{"cell_type":"markdown","source":["**What are the similarities between these trees? What are the differences?**"],"metadata":{"id":"JUJ1LfobhJgy"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"N7uXmyuLhOz5"}},{"cell_type":"markdown","source":["© Elisabeth Richardson, 2023. Adapted from Applied Bioinformatics by David A. Hendricks under a [CC-by-4.0 license](https://https://creativecommons.org/licenses/by/4.0/) and shared under same."],"metadata":{"id":"yk7P4f15hRN6"}}]}