{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOd9bk3EDfH/4x5KP7EB+tl"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**WEEK 4**\n","\n","This week we'll be working through an adapted version of Chapter 2 from Applied Bioinformatics.\n","\n","First, we have to install BioPython again, since this is a new notebook. You can do that by running the code block below."],"metadata":{"id":"Ndx2IJDmAL2Y"}},{"cell_type":"code","execution_count":1,"metadata":{"id":"Kzb9uvD4AHxL","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1702438505934,"user_tz":420,"elapsed":11904,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"cbd96b98-cfd8-431c-9d05-c2bdcaa32dd6"},"outputs":[{"output_type":"stream","name":"stdout","text":["Collecting biopython\n","  Downloading biopython-1.81-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m14.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.23.5)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.81\n"]}],"source":["##Install BioPython in Jupyter Notebook\n","%pip install biopython"]},{"cell_type":"markdown","source":["#Part 1: Introduction to Sequence Motifs.\n","First, make sure you've read Sections 2.1-3 of the textbook. If you haven't read it, do that now.\n","\n","The first examples from 2.1 and 2.2. involve finding a specific sequence within a longer DNA string. This is a function that is very common in bioinformatics and, therefore, built in to BioPython."],"metadata":{"id":"i1NNywv0Ac0I"}},{"cell_type":"code","source":["##Run the code below. What do the numbers mean and how are they defined?\n","from Bio.Seq import Seq\n","from Bio import SeqUtils\n","pattern = Seq(\"ACG\")\n","sequence = Seq(\"ATGCGCGACGGCGTGATCAGCTTATAGCCGTACGACTGCTGCAACGTGACTGAT\")\n","results = SeqUtils.nt_search(str(sequence),pattern)\n","print(results)"],"metadata":{"id":"5bPXTYd6PaOE"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["##The code below searches for a sequence in the reverse complement of a string. When might you want to search a reverse complement?\n","results_rc = SeqUtils.nt_search(str(sequence),pattern.reverse_complement())\n","print(results_rc)"],"metadata":{"id":"CwApYm9iQ-Gz"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now it's time for Exercise 6:\n",">Find the number of occurrences and locations of `ACTT` (including the forward and reverse strand) within the following sequence:\n",">\n",">```\n",">sequence\n","AGCGATCTAGCATACTTATACGCGCGCAGCTATCGATCACTTGTGCTAGTAAAGTGCGCGCCGCATTAAAGTGCTAGCTAGCTACTTAGCTAGCTAGTCG\n","```\n","\n"],"metadata":{"id":"5U3HjZ8pR0s9"}},{"cell_type":"code","source":["##You can write your code in this block."],"metadata":{"id":"rTJowln5RTS3"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#Part 2: Consensus Sequences\n","First, make sure you've read Sections 2.1-3 of the textbook. If you haven't read it, do that now.\n","\n","First: Why do we as biologists care about consensus sequences? Give an example of a consensus sequence. (If you've already taken BIOL3101 this should be very easy!)"],"metadata":{"id":"bilbxKSUSgLh"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"Pqfiq1fsTCuu"}},{"cell_type":"markdown","source":["As biologists care about consensus sequences a lot, as you can imagine there is a function built in to BioPython to search for them."],"metadata":{"id":"ZqtrDKRzTGe_"}},{"cell_type":"code","source":["##Run the code below. What does the resulting output mean?\n","from Bio import SeqUtils\n","consensus = \"RGWYV\"\n","sequence = \"CGTAGCTAGCTCAGAGCAGGGACACGTGCTAGCAACAGCGCT\"\n","results = SeqUtils.nt_search(sequence,consensus)\n","print(results)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"JhjTI8erTW9D","executionInfo":{"status":"ok","timestamp":1702439992000,"user_tz":420,"elapsed":199,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"741a50f7-39ad-4507-af9b-738edbfab126"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["['[AG]G[AT][CT][ACG]', 19]\n"]}]},{"cell_type":"markdown","source":["The concept of a consensus sequence is really important for when we're doing BLAST later, so also explain below in your own words what the following variables represent. Give at least three different sequences that satisfy the criteria:\n","1. consensus\n","2. results"],"metadata":{"id":"v6RnCYs5U7P2"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"sQxi_BuHVfl9"}},{"cell_type":"markdown","source":["#Part 3: Finding new motifs\n","Up until this point, we've been pre-defining our motifs and searching for them in a larger sequence. We already know a lot of motifs from our previous studies: for example, you might know the motif \"AUG\" as the start codon. However, this only works when we know exactly what we're looking for within the genome. Unfortunately, we *don't* always know exactly when a motif will occur in a genome, how many motifs there are, or what a motif looks like.\n","\n","We can calculate these things using a weight matrix; these basically summarise the probability that a given sequence in a dataset is a recurring motif rather than a random sequence. This is difficult, because you can only determine this in the context of a whole genome.\n","\n","Think about playing Snap. I could shuffle a deck of cards, get you to draw a card and ask you if it's Snap or not. You'd tell me, well that depends on the context of the other cards, what's ahead and what's before. It's kind of the same with motif searching.\n","\n","There is a LOT of math in this section of the textbook that is beyond the scope of this course, which is why I'm not suggesting you understand the entire section before you start this bit of the lab. However, I do encourage you to read it and ask Brian lots of hard questions.\n","\n","As before, finding motifs *de novo* (this means without additional info like what motif we're looking for) is very important in bioinformatics and there is a function for it in BioPython. Before we use that function, however, we have to learn a tiny bit more about how the Python programming language works."],"metadata":{"id":"HjLR2ncUV7V0"}},{"cell_type":"code","source":["##This code creates a variable called instances and saves a list of DNA sequences to it.\n","instances = [Seq(\"CAGTT\"),Seq(\"CATTT\"),Seq(\"ATTTA\"),Seq(\"CAGTA\"),Seq(\"CAGTT\"),Seq(\"CAGTA\")]"],"metadata":{"id":"GtWoUqzJeGjE","executionInfo":{"status":"ok","timestamp":1702442372790,"user_tz":420,"elapsed":191,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}}},"execution_count":15,"outputs":[]},{"cell_type":"markdown","source":["When you run the code above, you create a variable to which a value is assigned, just like before. However, up until this point we've been working with just numbers and strings: a string is a sequence of characters and usually used for any text, like `Hello world` or `ATGTCTGTGT` and is defined in python using quotation marks:"],"metadata":{"id":"qfMyTU0-eEFj"}},{"cell_type":"code","source":["x = \"hello world\"\n","print(x)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"grJhqPjkf2ai","executionInfo":{"status":"ok","timestamp":1702442798330,"user_tz":420,"elapsed":184,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"28f81eda-0354-44ce-f1d8-c9463fed1d0b"},"execution_count":16,"outputs":[{"output_type":"stream","name":"stdout","text":["hello world\n"]}]},{"cell_type":"markdown","source":["In this code, we're using a list. A list is an ordered collection of elements which can be manipulated with various Python functions, and unlike strings are not limited to characters. A list is defined in Python by square brackets, with the elements separated by commas."],"metadata":{"id":"ZgHPApMLf5dN"}},{"cell_type":"code","source":["x = [1, 'hello', 3.14, True]\n","print(x)\n","print(x[1])"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"pdSXa2UIgc5z","executionInfo":{"status":"ok","timestamp":1702443073945,"user_tz":420,"elapsed":292,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"44a0a1f1-13f5-4960-8459-da9d22ff68c1"},"execution_count":18,"outputs":[{"output_type":"stream","name":"stdout","text":["[1, 'hello', 3.14, True]\n","hello\n"]}]},{"cell_type":"markdown","source":["Create a list that includes five elements and at least three data types. Print the first, third, and fifth elements of the list."],"metadata":{"id":"R3bcm_Eni36n"}},{"cell_type":"code","source":["##You can write your code in this block."],"metadata":{"id":"D3C--2Yrg9v7"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now let's go back to our motif function and our `instances` list."],"metadata":{"id":"zzL1zECHhKCi"}},{"cell_type":"code","source":["#This code takes six short DNA strings (contained in the variable instances) and saves it as a motif.\n","from Bio import motifs\n","from Bio.Seq import Seq\n","motif = motifs.create(instances)\n","print(motif)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"2s_XSiwQcdz2","executionInfo":{"status":"ok","timestamp":1702443154224,"user_tz":420,"elapsed":183,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"cc37ade5-40b1-4975-d01f-979b2c0726c8"},"execution_count":19,"outputs":[{"output_type":"stream","name":"stdout","text":["CAGTT\n","CATTT\n","ATTTA\n","CAGTA\n","CAGTT\n","CAGTA\n","\n"]}]},{"cell_type":"markdown","source":["At first glance, this doesn't look very interesting - we already had a variable `instances` that could print our list of sequences. However, by creating a motif of this list, we can identify all sorts of other attributes. For example, we can use the code below to quantify the amount of variation at each site in our motif."],"metadata":{"id":"Js0jJo10dTSI"}},{"cell_type":"code","source":["##This code prints the motif counts of our sequences\n","print(motif.counts)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"VM0RmSrhdR_R","executionInfo":{"status":"ok","timestamp":1702442118382,"user_tz":420,"elapsed":215,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"07dfb226-ed6e-4928-904b-34cc1228ef03"},"execution_count":12,"outputs":[{"output_type":"stream","name":"stdout","text":["        0      1      2      3      4\n","A:   1.00   5.00   0.00   0.00   3.00\n","C:   5.00   0.00   0.00   0.00   0.00\n","G:   0.00   0.00   4.00   0.00   0.00\n","T:   0.00   1.00   2.00   6.00   3.00\n","\n"]}]},{"cell_type":"markdown","source":["We can also identify the consensus sequence."],"metadata":{"id":"FMwXwC37hulX"}},{"cell_type":"code","source":["##This code prints the motif consensus of our sequences\n","print(motif.degenerate_consensus)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"tlYa70xKhw4E","executionInfo":{"status":"ok","timestamp":1702443299450,"user_tz":420,"elapsed":267,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"6df6c7a3-aff0-408b-97b3-4c5a6a3f16d5"},"execution_count":20,"outputs":[{"output_type":"stream","name":"stdout","text":["CAKTW\n"]}]},{"cell_type":"markdown","source":["This code identifies the degenerate consensus sequence. What does the term degenerate mean in biology, and what does it mean in this context?"],"metadata":{"id":"Vmih1jbWh5uk"}},{"cell_type":"markdown","source":["\\---------------------------\n","You can add your answer in this text box.\n","\\---------------------------"],"metadata":{"id":"D9kG0DZWiDMf"}},{"cell_type":"markdown","source":["The motifs tool also allows us to visualize our motifs."],"metadata":{"id":"yiH8pFvwiMv8"}},{"cell_type":"code","source":["#This code creates an image of your motif.\n","motif.weblogo(\"LOGO.pdf\",format=\"pdf\")"],"metadata":{"id":"soEcG01giC4B","executionInfo":{"status":"ok","timestamp":1702443447098,"user_tz":420,"elapsed":1326,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}}},"execution_count":21,"outputs":[]},{"cell_type":"markdown","source":["You might recognise these extremely fancy diagrams from other genetics classes."],"metadata":{"id":"VmvlTs3nia7i"}},{"cell_type":"markdown","source":["© Elisabeth Richardson, 2023. Adapted from Applied Bioinformatics by David A. Hendricks under a [CC-by-4.0 license](https://https://creativecommons.org/licenses/by/4.0/) and shared under same."],"metadata":{"id":"f4933PDCip1D"}}]}