{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOpHIpE0iptJe7Z0TZUdEag"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["#**WEEK 2**\n","\n","This week, we'll be working through a slightly adapted version of Exercises 1-3 and Lab 1 from Chapter 1 of Applied Bioinformatics."],"metadata":{"id":"iYfzSnSHRqyP"}},{"cell_type":"markdown","source":["#Part 1: Your File Organization in Jupyter Notebooks\n","Look to the left of the screen and click on the little folder icon. That's your file directory structure within Jupyter Notebooks. You'll notice there's a file in there called \"sequences.fa\" - we'll use that later. There's also a folder there called \"sample_data\" which we will not use later but it appears by default in this notebook setup and there is a 100% chance I will forget to delete it. Feel free to look at the files if you want to know data about the California housing market."],"metadata":{"id":"wvEhxjJdRrxL"}},{"cell_type":"code","source":["##You can edit your directory structure by using bash commands in one of these code blocks. Try running this code block, then check your directory again.\n","\n","!mkdir new_directory\n","\n","##Try making your own new directory!\n","\n"],"metadata":{"id":"5aPqtMbATwPe"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Did you notice the difference between this command and the commands you used in bashcrawl? This one has an exclamation point in front of it, because that's a signal to the Jupyter Notebook interface to tell it how to interpret the command. Since this is an interactive Python notebook (or .ipynb) then it defaults to trying to use Python first, while your terminal will default to using Bash first."],"metadata":{"id":"gFDtcB9nU8o4"}},{"cell_type":"code","source":["##Take some time here to practice navigating the directory structure using terminal commands.\n","##Below, show how you would make a new directory called \"scripts\".\n","##Move into it, and create a file called \"test_script.py\" that prints \"Hello world!\"\n","##You can use the instructions in Chapter 1 of Applied Bioinformatics to help you."],"metadata":{"id":"VNY3uWjNGoNY"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#Part 2.1: Sequences, Strings and BioPython\n","\n","We're going to be working through the content from here of Part 1.2 of Applied Bioinformatics. If you've not read the section, read it now."],"metadata":{"id":"xADANngqX1GF"}},{"cell_type":"markdown","source":["Now, let's try some of the examples. To do this, we need to install Biopython in the Jupyter notebook environment. You can do this by running the block of code below. It might take a minute or two. Also, because of how Jupyter notebook works you may have to reinstall Biopython any time you close the current notbook session: if you start seeing warning signs that look like this:\n","\n","\n","\n","```\n","ImportError: cannot import <whatever BioPython command you were trying to use>\n","```\n","If you get those, just run the code block below this message again.\n"],"metadata":{"id":"ziWY9GIVYco-"}},{"cell_type":"code","source":["##Install BioPython in Jupyter Notebook\n","%pip install biopython"],"metadata":{"id":"c2YH32A4ZwLg"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, let's look at some sequence data."],"metadata":{"id":"M95R9oSvax8l"}},{"cell_type":"code","source":["#A DNA sequence is just a string of characters that can be easily manipulated by a computer. We can work with strings in base Python.\n","\n","x = \"ATGGCGGGAAAATGA\"\n","\n","print(x)\n","\n","print(x[5])\n","\n","#Now try printing the first T\n","\n","#Now try printing the last G\n"],"metadata":{"id":"3rtxsF-EZPrK"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["##We can also work with DNA sequence strings in BioPython.\n","##This may look like more work now, but when you're working with real sequence data it will be MUCH easier.\n","\n","from Bio.Seq import Seq\n","x = Seq(\"ATGGCGGGAAAATGA\")\n","Seq('ATGGCGGGAAAATGA')\n","x[5]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":35},"id":"3zwbdsffa48s","executionInfo":{"status":"ok","timestamp":1701418194876,"user_tz":420,"elapsed":248,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"de4fc626-f4a9-4887-db14-5b4b04e2f1a4"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'G'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":29}]},{"cell_type":"markdown","source":["You might have also noticed that this command looks slightly different to how it is written in the textbook. This is because BioPython did a giant update over the summer which is the current bane of my (Beth's) existence because it broke a whole bunch of software I use for research and I am currently begging people to fix it for free because this is how academia works. This is also why some of the commands in the textbook are also broken. You get what you pay for. Assume the commands in the textbook are correct unless it's different here."],"metadata":{"id":"DkQu5aOmbjZq"}},{"cell_type":"code","source":["##Let's try transcribing a sequence\n","from Bio.Seq import Seq\n","x = Seq(\"ATGGCGGGAAAATGA\")\n","print(x.transcribe())\n","\n","##Let's try translating a sequence\n","print(x.translate())\n","\n","##Let's try measuring the GC content of a sequence\n","from Bio.SeqUtils import gc_fraction\n","print(f\"GC content of {x} : {gc_fraction(x):.2f}\")\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Q4x0eCuJb7ZF","executionInfo":{"status":"ok","timestamp":1701463017018,"user_tz":420,"elapsed":198,"user":{"displayName":"Beth Richardson","userId":"17739406692480062480"}},"outputId":"08c25352-8208-458d-87ff-0bc77cd448e9"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["AUGGCGGGAAAAUGA\n","MAGK*\n","GC content of ATGGCGGGAAAATGA : 0.47\n"]}]},{"cell_type":"markdown","source":["So as well as just defining a DNA sequence as a string, using BioPython allows us to carry out biological operations like transcribing and translating DNA. You can also assign these items to different variables - instead of printing the transcription of x, you can save it as x.RNA.\n","\n","\n","\n","```\n","x = Seq(\"ATGGCGGGAAAATGA\")\n","\n","x.RNA = x.transcribe()\n","```\n","\n","Try completing Exercise 2 from Applied Bioinformatics below. Don't forget to import BioPython!\n","\n",">Exercise 1\n",">\n",">Transcribe the following DNA sequence and translate the resulting mRNA sequence. What protein sequence do you get? Assign each sequence to its own variable and then print them to the console.\n",">\n",">\\>CDS ATGCTGCGGCGAGCTCCGTCCAAAAGAAAATGGGGTTTGGTGTAAATCTGGGGGTGTAATGTTATCATATAAATAAAAGAAAATGTAAAAAACAAAAACAAAAACAAAAAAGCC\n"],"metadata":{"id":"AfGPw_SJFqZs"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"L-xN5NI_Hvdu"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#Part 2.2: Subsequences and Reverse Complement\n","We're now at Section 1.2.3, Subsequences and Reverse Complement, and the training wheels are going to start to come off a little bit. Using the explanations given in the textbook and the commands you've learned, try to print the sequence that corresponds to the first three codons of CDS in the next codeblock. Don't forget that the Alphabet parameter is depcreciated, so we don't have to include that."],"metadata":{"id":"iwaqVAOBJDZB"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"2CwUGZRfJrmx"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now try to produce the reverse complement of CDS."],"metadata":{"id":"U6IB7U8YJ5mK"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"oTI-IrBPJ8sL"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now we're going to try Exercise 2 using the information you've learned so far.\n","\n",">Exercise 2\n",">You can concatenate two sequences (attach them to each other to form one longer sequence) with the \"+\" operator, such as  \n",">\n",">`D = D1 + D2 `\n",">\n",">How can you use this to create a very long palindromic sequence?"],"metadata":{"id":"Z619WPPdKdId"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"kXPA5WtAK3Wl"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["#Part 3: File Formats\n","This section corresponds to Section 1.3 of Applied Bioinformatics. If you haven't read it, do it now. Surprise! We'll be using our sequences.fa file now.\n","\n","Try viewing the file in the command line. Don't forget how to format bash commands in Jupyter notebooks.\n","\n"],"metadata":{"id":"90qcxkyVK7XU"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"0XuakGnMM5lA"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now let's try importing and printing our sequences.fa file using the for loop structure discussed in the texbook."],"metadata":{"id":"D2_S3_sNQ17U"}},{"cell_type":"code","source":["from Bio import SeqIO\n","\n","fastaFile = \"sequences.fa\"\n","sequences = SeqIO.parse(open(fastaFile),'fasta')\n","\n","for record in sequences:\n","    print(record.id, record.seq)"],"metadata":{"id":"0o7tyidpQ_yB"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Now, let's try Exercise 3 from the textbook. You can use the given \"sequences.fa\" file or make a new fasta file.\n","\n",">Exercise 3\n",">\n",">Write a Biopython script that reads in a FASTA file, and prints a new FASTA file with the reverse complement of each sequence.\n",">\n",">Write a script to read a FASTA file and print the reverse complement of each sequence. Print the GC content of each sequence."],"metadata":{"id":"or-sYryIRf2T"}},{"cell_type":"code","source":["##Your code here"],"metadata":{"id":"2IBg4u8pSLUw"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["Â© Elisabeth Richardson, 2023. Adapted from Applied Bioinformatics by David A. Hendricks under a [CC-by-4.0 license](https://https://creativecommons.org/licenses/by/4.0/) and shared under same."],"metadata":{"id":"Gt6bbWz0UJl_"}}]}